# haplotype network analysis with R and the ape and pegas packages
#install.packages("devtools", dependencies = TRUE)
#library(devtools)
#install.packages("pegas", dependencies = T)
#install_github("thibautjombart/adegenet")
#install_github("emmanuelparadis/pegas/pegas")
library(pegas) #ape v5.3 and pegas v0.13
library(DECIPHER)
library(magrittr) 
library(dplyr)
setwd("~/Documents/Msc_Bioinformatics_Project/IgDiscover/Networks/")

#### ALIGN SEQUENCES FIRST ####
seqs <- readDNAStringSet("./combined_V_germline_filter.fasta", format = "fasta")
# nucleotide sequences need to be in the same orientation
# if they are not, then they can be reoriented (optional)
seqs <- OrientNucleotides(seqs)
# perform the alignment
aligned <- AlignSeqs(seqs)
# write the alignment to a new FASTA file
writeXStringSet(aligned,file="filter.fasta")

#### READING THE INPUT DATA (ALIGNMENT) ####
# We first read the alignment into a DNAbin object:
alignment <- read.dna("filter.fasta", format = "fasta") 
class(alignment) # it's an object of class "DNAbin"
str(alignment) # 45 sequences

#### CREATE A HAPLOTYPE NETWORK ######
h <- haplotype(alignment)

# create a table with sequence name and haplotype names 
## this table will have 45 rows:
haplotype_t <- tibble(sequence_name = character(), haplotype_name = character())
#concatenate sequence name and the type of haplotype 
name <- paste(rownames(alignment), rownames(h), sep = " - ")

haplotype_t %<>%bind_rows(list(sequence_name = name, haplotype_name = rownames(h)))
haplotype_table <- table(haplotype_t)
nrow(haplotype_table) # ok, 45

#check out just a chuck of the table 1-10 row and column 
haplotype_table[1:10, 1:10]  
#create a network
net <- haploNet(h)
#plot the network
pdf("Network.pdf", width = 14, height = 10)
plot(net, size= 5, pie = haplotype_table, scale.ratio = 1, cex = 0.4, threshold = 0, 
     show.mutation = 0, asp = 1, fast = F, legend=FALSE, main = "Individualized V gene database generated by IgDiscover")
legend(x = -150, y = 100, ncol = 1, legend = rownames(haplotype_table),
       fill = rainbow(nrow(haplotype_table)), bty = "n", pch = 19, title = "V alleles")
dev.off()

#exploring interacting with netwworkD3 just incase haplotype networks don't work ##########################
library(networkD3)



